File: evm/python/examples/create_wallet.py
"""Create a wallet controlled by the caller and fund its vault."""

import json
import os

from eth_account import Account
from web3 import Web3
from web3.middleware import SignAndSendRawMiddlewareBuilder

from swig_evm import get_contract
from swig_evm.authorities import Secp256k1, encode_authority
from swig_evm.permissions import SimplePermission, encode_permission

web3 = Web3(Web3.HTTPProvider(os.environ["RPC_URL"]))
signer = Account.from_key(os.environ["SIGNER_PRIVATE_KEY"])
web3.middleware_onion.inject(SignAndSendRawMiddlewareBuilder.build(signer), layer=0)
web3.eth.default_account = signer.address
factory = get_contract(web3, "SwigConfigFactory", os.environ["FACTORY_ADDRESS"])
# The name is public. Identical name/root settings predict the same wallet.
wallet_id = Web3.keccak(text=os.environ["WALLET_NAME"])
args = (
    wallet_id,
    wallet_id,
    *encode_authority(Secp256k1(signer.address)),
    [encode_permission(SimplePermission.ALL)],
)
predicted = factory.functions.computeConfigAddress(*args).call()
hash_value = factory.functions.deploy(*args).transact(
    {"value": int(os.environ["INITIAL_BALANCE_WEI"])}
)
receipt = web3.eth.wait_for_transaction_receipt(hash_value)
if receipt["status"] != 1:
    raise RuntimeError("Wallet creation reverted")
# Decode only the factory's event; initialization also emits account events.
log = next(log for log in receipt["logs"] if log["address"] == factory.address)
deployment = factory.events.SwigDeployed().process_log(log)["args"]
if deployment["config"] != predicted:
    raise RuntimeError("Deployed wallet differs from prediction")
print(
    json.dumps(
        {
            "accountAddress": deployment["config"],
            "vaultAddress": deployment["vault"],
            "capsuleAddress": deployment["capsule"],
            "transactionHash": Web3.to_hex(hash_value),
        }
    )
)

File: evm/rust/examples/create_wallet.rs
//! Create a wallet controlled by the caller and fund its vault.
use alloy::primitives::keccak256;
use alloy::{
    network::EthereumWallet,
    primitives::{Address, U256},
    providers::ProviderBuilder,
    signers::local::PrivateKeySigner,
};
use std::{env, error::Error};
use swig_evm::{
    authorities::Authority, contracts::factory::SwigConfigFactory, permissions::Permission,
};

#[tokio::main]
async fn main() -> Result<(), Box<dyn Error>> {
    let signer: PrivateKeySigner = env::var("SIGNER_PRIVATE_KEY")?
        .parse()
        .map_err(|_| "invalid signing key")?;
    let owner = signer.address();
    let provider = ProviderBuilder::new()
        .wallet(EthereumWallet::from(signer))
        .connect_http(env::var("RPC_URL")?.parse()?);
    let address: Address = env::var("FACTORY_ADDRESS")?.parse()?;
    let factory = SwigConfigFactory::new(address, provider);
    let authority = Authority::Secp256k1 { address: owner }.encode()?;
    // The public name and root settings determine the address; choose a fresh name.
    let id = keccak256(env::var("WALLET_NAME")?);
    let actions = vec![Permission::All.encode()?.into()];
    let predicted = factory
        .computeConfigAddress(
            id,
            id,
            authority.authorityType,
            authority.key,
            authority.keyExtra,
            actions.clone(),
        )
        .call()
        .await?;
    let receipt = factory
        .deploy_call(
            id,
            id,
            authority.authorityType,
            authority.key,
            authority.keyExtra,
            actions,
        )
        .value(env::var("INITIAL_BALANCE_WEI")?.parse::<U256>()?)
        .send()
        .await?
        .get_receipt()
        .await?;
    if !receipt.status() {
        return Err("wallet creation reverted".into());
    }
    let event = receipt
        .decoded_log::<SwigConfigFactory::SwigDeployed>()
        .ok_or("missing wallet deployment event")?;
    if event.address != address || event.data.config != predicted {
        return Err("deployment event differs from prediction".into());
    }
    println!(
        "{}",
        serde_json::json!({"accountAddress": event.data.config,
        "vaultAddress": event.data.vault, "capsuleAddress": event.data.capsule,
        "transactionHash": receipt.transaction_hash})
    );
    Ok(())
}
